src/theory/datatypes/inference_manager.h

```diff
@@ -0,0 +1,111 @@
+/*********************                                                        */
+/*! \file inference_manager.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2020 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Datatypes inference manager
+ **/
+
+#include "cvc4_private.h"
+
+#ifndef CVC4__THEORY__DATATYPES__INFERENCE_MANAGER_H
+#define CVC4__THEORY__DATATYPES__INFERENCE_MANAGER_H
+
+#include "context/cdhashmap.h"
+#include "expr/node.h"
+#include "theory/inference_manager_buffered.h"
+
+namespace CVC4 {
+namespace theory {
+namespace datatypes {
+
+/**
+ * The datatypes inference manager. The main unique features of this inference
+ * manager are:
+ * (1) Explicit caching of lemmas,
+ * (2) A custom process() method with relies on a policy determining which
+ * facts must be sent as lemmas (mustCommunicateFact).
+ * (3) Methods for tracking when lemmas and facts have been processed.
+ */
+class InferenceManager : public InferenceManagerBuffered
```

Comment:
I really don't like how this class uses `InferenceManagerBuffered`. The pure existence of `doSendLemma()` suggests to me that one is not supposed to call `addPendingLemma()`, though this is technically allowed (by `public` inheritance) and also expected (`process()` expects some lemmas to be pending as it calls `doPendingLemmas()`). This means that certain lemmas are cached, and some are not, without documentation, though.

Reply:
Discussed offline, I am standardizing both the caching of lemmas and the tracking of whether lemmas/facts have been sent, thus eliminating the need for doSendLemma.

Reply:
I like the new solution way better.